### PR TITLE
feat: mixture samples in improved reaction scheme

### PR DESCRIPTION
**Pull request by TasnimMehzabin**
* feat: enhance component creation and sample handling for mixtures in the reaction scheme

- Added logic to pre-calculate relative molecular weights for mixture samples before saving.
- Introduced bulk molecule data preloading for components to optimize queries.
- Updated `Components::Create` use case to process `Sample` objects directly instead of IDs.
- Adjusted `Reactable` behavior to skip updates for mixture samples.
- Removed redundant molecule association logic from `ComponentAPI`.
- Improved handling of components in reaction material updates and sample modifications.

* feat: improve UI styling and refine handling of mixture components

- Enhanced `Components` accordion styling and accessibility for better user experience.
- Added `ReactionMaterialComponentsGroup` for structured display of mixture components.
- Centralized `isMixture` logic and updated state management in `Material.js`.
- Refactored target amount and total mass calculations to account for liquid component masses.
- Improved component deserialization and reaction material updates in `ReactionDetailsScheme.js`.

* feat: add volume and concentration flag to reactions

- Introduced `volume` and `use_reaction_volume_for_concentration` to `reactions` table.
- Updated entities and API to handle the new fields.
- Added migrations for schema changes.

* feat: streamline component handling and improve sample workflows

- Simplified component operations using helper methods (`normalizeComponents`, `findComponentById`).
- Enhanced handling of locking mechanisms with per-sample states for amount/solids columns and reaction equivalence.
- Refactored state logic for locking and concentration calculations in `ComponentStore`.
- Improved UI and calculation logic for mixture mass, volume, and concentration.
- Declared `setComponentDeletionLoading` as a required prop and improved deletion state handling.

* test: remove redundant `referenceComponent` parameter from methods and update tests

- Simplified methods in `Component` by eliminating `referenceComponent` parameter.
- Adjusted associated test cases to reflect updated method signatures.
- Expanded test coverage for volume and concentration edge cases in `Reaction` and `Sample` classes.
- Enhanced `Reaction` and `API` tests to incorporate `volume` and `use_reaction_volume_for_concentration` attributes.
- Improved concentration calculations and mass handling for mixtures and reactions with shared state.

* test: extend component and sample test coverage with serialization improvements

- Expanded `Component` and `Sample` class tests to cover edge cases and property calculations.
- Added serialization tests for `Component` to ensure accurate data handling.
- Improved validation, error handling, and molecular weight calculations in `Components::Create` use case.

### CODE REVIEW

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/Material.js [hunk 1]
```diff
@@ -379,24 +436,205 @@ class Material extends Component {
...
+      this.setState({
+        mixtureComponents: componentsList,
+        mixtureComponentsLoading: false,
+      });
+    } else if (typeof material.id === 'number') {
+      // Fetch components for saved material
+      this.setState({ mixtureComponentsLoading: true });
+
+      ComponentsFetcher.fetchComponentsBySampleId(material.id)
+      .then((components) => {
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/Material.js]: for this block: Expected indentation of 10 spaces but found 8.eslint[indent](https://eslint.org/docs/latest/rules/indent)

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/Material.js [hunk 2]
```diff
@@ -743,21 +964,61 @@ class Material extends Component {
...
+        {isMixture && hasComponents && (
+          <Accordion
+            className="mixture-components-accordion"
+            activeKey={showComponents ? 'components' : null}
+            onSelect={this.toggleComponentsAccordion}
+          >
+            <Accordion.Item eventKey="components">
+              <Accordion.Header>Components</Accordion.Header>
+              <Accordion.Body>
+                <div className="mixture-components-row" style={{ padding: 0, background: '#f8f9fa' }}>
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/Material.js]: would be better if inline styling can be avoided here

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js [hunk 3]
```diff
@@ -103,33 +137,77 @@ export default class ReactionDetailsScheme extends React.Component {
...
-            splitSample.target_amount_value = comp.amount_mol / comp.molarity_value;
-            splitSample.target_amount_unit = 'l';
-          }
-          reaction.addMaterialAt(splitSample, null, tagMaterial, tagGroup);
-          onReactionChange(reaction, { updateGraphic: true });
-        })
-        .catch((errorMessage) => {
-          console.log(errorMessage);
-        });
+      this.handleMixtureSample(splitSample, srcSample, reaction, tagMaterial, tagGroup)
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js]: missing ;

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js [hunk 4]
```diff
@@ -103,33 +137,77 @@ export default class ReactionDetailsScheme extends React.Component {
...
+    const sampleComponents = this.copyComponents(srcSample.components, splitSample.id);
+    splitSample.components = sampleComponents;
+
+    this.setTargetAmountFromComponents(splitSample, sampleComponents);
+    this.addSampleToReaction(splitSample, srcSample, reaction, tagMaterial, tagGroup);
+  }
+
+  handleMixtureWithApiComponents(splitSample, srcSample, reaction, tagMaterial, tagGroup) {
+    ComponentsFetcher.fetchComponentsBySampleId(srcSample.id)
+    .then(async (components) => {
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js]: Expected indentation of 6 spaces but found 4.eslint[indent](https://eslint.org/docs/latest/rules/indent)

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js [hunk 5]
```diff
@@ -103,33 +137,77 @@ export default class ReactionDetailsScheme extends React.Component {
...
+    });
+  }
+
+  setTargetAmountFromComponents(splitSample, sampleComponents) {
+    const validComponent = sampleComponents.find(
+      (component) => component.amount_mol > 0 && component.molarity_value > 0
+    );
+
+    if (validComponent) {
+      splitSample.target_amount_value = validComponent.amount_mol / validComponent.molarity_value;
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js]: is there always a valid value for molarity_value? - seems like a try, catch block here might be helpful

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js [hunk 6]
```diff
@@ -430,15 +514,37 @@ export default class ReactionDetailsScheme extends React.Component {
...
     const { sampleID, amount } = changeEvent;
     const updatedSample = this.props.reaction.sampleById(sampleID);
+
+    // --- Validate mixture mass before applying the change ---
+    const exceedsMassLimit = updatedSample.validateMixtureMass(amount);
+    if (exceedsMassLimit) {
+      const totalMixtureMass = updatedSample?.total_mixture_mass_g;
+      NotificationActions.add({
+        title: 'Mass Exceeded',
+        message: `Entered mass exceeds the available mixture mass for this sample. ` +
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js]: Strings must use singlequote.eslint[quotes](https://eslint.org/docs/latest/rules/quotes)

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js [hunk 7]
```diff
@@ -639,6 +745,91 @@ export default class ReactionDetailsScheme extends React.Component {
...
+    }
+
+    // Get the parent sample amount in mol for validation
+    const parentAmountMol = updatedSample.amount_mol;
+    if (parentAmountMol <= 0) {
+      console.warn('Sample amount (mol) is zero or negative, cannot calculate properties');
+      return;
+    }
+
+    // updatedSample.calculateMassFromReferenceComponent(referenceComponent);
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js]: delete

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js [hunk 8]
```diff
@@ -865,6 +1056,32 @@ export default class ReactionDetailsScheme extends React.Component {
...
   }
 
+  // eslint-disable-next-line class-methods-use-this
+  // Handle mixture samples differently
+  handleEquivalentBasedAmountUpdate(sample, newAmountMol) {
+    if (sample.isMixture() && sample.hasComponents()) {
+      // For mixture samples, we need to use the reference component's relative molecular weight
+      // to calculate the correct mass from the new amount_mol
+      const referenceComponent = sample.reference_component;
+      if (referenceComponent && referenceComponent.relative_molecular_weight) {
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js]: move this if block to parent if block and delete nested else, since it is the same else block of parent if-else block

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js [hunk 9]
```diff
@@ -865,6 +1056,32 @@ export default class ReactionDetailsScheme extends React.Component {
...
+        const newAmountG = newAmountMol * referenceComponent.relative_molecular_weight;
+        sample.setAmount({ value: newAmountG, unit: 'g' });
+      } else {
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionDetailsScheme.js]: else block not needed

File: app/javascript/src/models/Sample.js [hunk 10]
```diff
@@ -641,6 +636,114 @@ export default class Sample extends Element {
...
+  updateConcentrationFromSolvent(reaction) {
+    const combinedVolume = reaction ? reaction.calculateCombinedReactionVolume() : null;
+
+    if (!combinedVolume || combinedVolume <= 0) {
+      this.concn = null;
+      return;
+    }
+
+    // Handle mixtures
+    if (this.isMixture && typeof this.isMixture === 'function' ? this.isMixture() : this.isMixture) {
```
Comment (adambasha0) [inline app/javascript/src/models/Sample.js]: not sure I understand this nested if condition, what is the difference between this.isMixture() and this.isMixture? the condition looks redundant

File: app/javascript/src/models/Sample.js [hunk 11]
```diff
@@ -657,24 +760,53 @@ export default class Sample extends Element {
...
+   * @private
+   * @param {Object} amount - The amount object containing value and unit
+   * @returns {void}
+   */
+  notifyReactionOfVolumeChange(amount) {
+    // Only notify if this sample belongs to a reaction and the unit is liters
+    if (this.belongTo && this.belongTo.type === 'reaction' && amount.unit === 'l') {
+      const reaction = this.belongTo;
+      if (typeof reaction.updateAllConcentrations === 'function') {
+        reaction.updateAllConcentrations();
```
Comment (adambasha0) [inline app/javascript/src/models/Sample.js]: the original block calculated solventVolume only for reactants and startingMaterials
On the other hand, updateAllConcentrations calculates solventVolume also for product Materials, is this intented? - just to make sure the logic here is correct

File: app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionMaterialComponentsGroup.js [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        <td>
+          <div
+            style={{ display: 'inline-block', maxWidth: '100%' }}>
+            <span style={{
+              display: 'block',
+              whiteSpace: 'nowrap',
+              overflow: 'hidden',
+              textOverflow: 'ellipsis',
+              maxWidth: '100%'
+            }}>
```
Comment (adambasha0) [inline app/javascript/src/apps/mydb/elements/details/reactions/schemeTab/ReactionMaterialComponentsGroup.js]: The closing bracket must be aligned with the line containing the opening tag (expected column 13 on the next line)eslint[react/jsx-closing-bracket-location](https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-closing-bracket-location.md)

File: app/javascript/src/models/Sample.js [hunk 13]
```diff
@@ -641,6 +636,114 @@ export default class Sample extends Element {
...
+   * No updates occur if the mixture or reference data are invalid.
+   *
+   * @method updateMixtureComponentAmounts
+   * @memberof Sample
+   *
+   * @example
+   * // Assuming `sample` is a mixture with components
+   * sample.amount_g = 10;
+   * sample.updateMixtureComponentAmounts();
+   * console.log(sample.components[0].amount_mol); // updated molar amount
```
Comment (adambasha0) [inline app/javascript/src/models/Sample.js]: sample.components[0].amount_mol

File: app/javascript/src/models/Sample.js [hunk 14]
```diff
@@ -657,24 +760,53 @@ export default class Sample extends Element {
...
+  /**
+   * Captures the previous molar and mass values before updating the amount,
+   * to support ratio-based recalculations in mixture mass conversions.
+   *
+   * @private
+   * @returns {void}
+   */
+  storePreviousAmountState() {
+    // Initialize the sample_details container if missing
+    this.initializeSampleDetails?.();
```
Comment (adambasha0) [inline app/javascript/src/models/Sample.js:809]: seems to be syntax error here

File: app/javascript/src/stores/alt/actions/ElementActions.js [hunk 15]
```diff
@@ -393,6 +394,9 @@ class ElementActions {
       SamplesFetcher.create(params)
         .then((result) => {
           dispatch({ element: result, closeView, components: params.components })
+        }).catch((errorMessage) => {
+          console.log(errorMessage);
+          LoadingActions.stop();
```
Comment (adambasha0) [inline app/javascript/src/stores/alt/actions/ElementActions.js:414]: what happens when it stops loading?

File: spec/javascripts/apps/mydb/elements/details/samples/propertiesTab/SampleComponent.spec.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
+const expect = require('expect');
+const React = require('react');
+const Sample = require('src/models/Sample').default;
+
+describe('SampleComponent', () => {
+  it('can be imported without errors', () => {
+    // Simple test to verify the component can be loaded
+    // More complex tests would require extensive mocking of stores/actions
+    expect(true).toBe(true);
```
Comment (adambasha0) [inline spec/javascripts/apps/mydb/elements/details/samples/propertiesTab/SampleComponent.spec.js]: redundant test

File: spec/javascripts/apps/mydb/elements/details/samples/propertiesTab/SampleComponent.spec.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  });
+
+  it('has expected Sample model functionality', () => {
+    const sample = new Sample({ id: 1, sample_type: 'mixture' });
+    expect(sample.id).toBe(1);
+    expect(sample.sample_type).toBe('mixture');
```
Comment (adambasha0) [inline spec/javascripts/apps/mydb/elements/details/samples/propertiesTab/SampleComponent.spec.js]: same

**Review by adambasha0 — CHANGES_REQUESTED** [review]

File: app/javascript/src/models/Sample.js (hunk 14 above)
Comment (TasnimMehzabin) [reply app/javascript/src/models/Sample.js:809]: That’s using the optional chaining operator (?.) to safely call a function only if it exists.
